turbonomic/prometurbo
# [OM-96115] Support prometurbo configuration through custom resource

## Intention

This PR allows `prometurbo` to obtain server and query configuration from custom resource defined in [turbo-metrics](https://pkg.go.dev/github.com/turbonomic/turbo-metrics@v0.0.0-20230222215340-3cdff28ffdaf/api/v1alpha1).

## Background
Currently `prometurbo` obtains server and query configuration from configMap, which is very difficult to edit. We decided to follow the custom resource pattern and allow users to define customer resources that describe prometheus servers, and prometheus query mappings. Users can change these resources any time, and `prometurbo` can discover them during each discovery.

## Implementation
Because `prometurbo` packaged in OVA still obtain configuration from configMap, we need to support both configMap and custom resource. To achieve this:

* Create the following interface 
```
type MetricProvider interface {
	GetTasks() (tasks []*Task)
}

```
* Move the existing configMap based implementation into `configmap.MetricProviderImpl`
* Create the new custom resource based implementation under `customresource.MetricProviderImpl`
* A large amount of refactoring is done to make the code easier to understand
* A [separate PR](https://github.com/turbonomic/prometurbo/pull/89) is created to handle installation, and operator. We now need to create `serviceaccount`, `clusterrole` and `clusterrolebinding` with permission to access custom API resources.

### TODO
* We need to append a clusterId to the IP address as the new stitching attribute. This will be done after `kubeturbo` side of changes are committed, otherwise stitching will fail in the platform.
* The `status` of the custom resources will be implemented separately.
* The multi-cluster support is not implemented as this is not required by customer, and there is no environment for testing. However it is easy to extend when needed.

### Testing
* Unit tests added for the new provider
* Deployed dev build in the test cluster with proper sa, clusterrole and clusterrolebinding
* Install CRD and create sample resources
* Observe the `prometurbo` log:
```
I0310 01:19:04.262641 1 prometurbo.go:108] Number of concurrent workers to discover metrics: 4
I0310 01:19:04.262854 1 server.go:73] HTTP server listens on: 10.254.16.95:8081
I0310 01:19:32.261532 1 provider.go:60] Discovered 1 PrometheusQueryMapping resources.
I0310 01:19:32.265387 1 provider.go:71] Discovered 1 PrometheusServerConfig resources.
I0310 01:19:32.265459 1 serverconfig.go:19] Loading PrometheusServerConfig turbo-community/prometheusserverconfig-singlecluster.
I0310 01:19:32.265469 1 client.go:68] Creating client for Prometheus server: http://prometheus.istio-system:9090
I0310 01:19:32.265473 1 serverconfig.go:35] There are 1 PrometheusQueryMapping resources in namespace turbo-community
I0310 01:19:32.265526 1 handler.go:158] Total discovery tasks to dispatch 1.
I0310 01:19:32.265561 1 dispatcher.go:44] worker 1 has received a task.
I0310 01:19:32.276051 1 dispatcher.go:47] worker 1 has finished.
I0310 01:19:32.276082 1 collector.go:38] Collected results from all 1 tasks.
I0310 01:19:32.276089 1 handler.go:167] Discovered 4 entities.
I0310 01:19:32.276335 1 topology.go:45] Created 1 services for twitter-cass-api [demoapp]
I0310 01:19:32.276346 1 topology.go:45] Created 1 services for twitter-cass-friend [demoapp]
I0310 01:19:32.276350 1 topology.go:45] Created 1 services for twitter-cass-tweet [demoapp]
I0310 01:19:32.276354 1 topology.go:45] Created 1 services for twitter-cass-user [demoapp]
```
* Verify the `prometurbo` output:
  
  <img width="479" alt="image" src="https://user-images.githubusercontent.com/10012486/224202177-0b0b8545-dfea-49d3-b7eb-346d1c515284.png">
